### PR TITLE
Provide a way to convert an exception into an RpcResponse in Thrift service

### LINKED ISSUE #3349: Provide a way to customize the exception in Thrift service.
In thrift services, when an exception is raised, a user might want to:
- Return a different exception
  - (e.g.) A user might want to convert an `IllegalArguementException` from Armeria or the 3rd party library to their own `BadRequestException`.
- Add some logic such as adding an HTTP header:
```java
.decorate(delegate -> new SimpleDecoratingRpcService(delegate) {
 @Override
 public RpcResponse serve(
         ServiceRequestContext ctx, RpcRequest req) throws Exception {
     final RpcResponse res = unwrap().serve(ctx, req);
     return RpcResponse.from(res.handle((obj, cause) -> {
         if (cause instanceof MyException) {
             ctx.addAdditionalResponseHeader(RETRY_AFTER, 3);
         }
         ...
     }));
 }
})
```
If we provide a way to customize an exception a user can achieve those easily:
```java
THttpService.builder()
            .addService("hello", helloService)
            .exceptionConverter((ctx, cause) -> {
                ctx.addAdditionalResponseHeader(RETRY_AFTER, 3); 
                return otherException; // We can also return another exception.
            })
```

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
A user might define his/her own exception in the Thrift service and use it in the client and server side.
If we provide a way to convert an exception into an `RpcResponse`, he/she can easily
translate the exceptions that are defined in JDK or 3rd party libraries into an `RpcResponse`.
Here are related resources.
- https://github.com/line/centraldogma/blob/master/server/src/main/java/com/linecorp/centraldogma/server/internal/thrift/CentralDogmaExceptionTranslator.java
- https://github.com/line/centraldogma/blob/master/server/src/main/java/com/linecorp/centraldogma/server/internal/thrift/Converter.java#L183
- https://github.com/line/centraldogma/blob/master/common-legacy/src/main/thrift/CentralDogma.thrift#L81

Modifications:
- Add `THttpServiceBuilder.exceptionHandler(...)`
- Clean up `THttpService`

Result:
- You can now convert an exception into an `RpcResponse` in `THttpService`.
  ```java
  THttpService.builder()
              .addService("hello", helloService)
              .exceptionHandler((ctx, cause) -> {
                  if (cause instanceof IllegalArgumentException) {
                      return RpcResponse.of(CustomizedException("Bad Request!"));
                  }
                  ...
              })
  ```
- Close #3349

### CODE REVIEW

File: thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java [hunk 1]
```diff
@@ -266,36 +258,25 @@ public static THttpService ofFormats(
...
+                        .otherSerializationFormats(otherSupportedSerializationFormats)
+                        .newDecorator();
     }
 
-    private final SerializationFormat[] supportedSerializationFormatArray;
-    private final Set<SerializationFormat> supportedSerializationFormats;
     private final ThriftCallService thriftService;
+    private final SerializationFormat defaultSerializationFormat;
+    private final Set<SerializationFormat> supportedSerializationFormats;
+    private final BiFunction<ServiceRequestContext, ? super Throwable, ? extends Throwable> exceptionMapper;
```
Comment (ikhoon) [inline thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java:269]: Like we did in others
https://github.com/line/armeria/blob/bd7fe49f6999af1d1c26b19dd7a220008c7ea202/core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java#L106
```suggestion
    private final BiFunction<? super ServiceRequestContext, ? super Throwable, ? extends Throwable> exceptionMapper;
```

File: thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java [hunk 2]
```diff
@@ -266,36 +258,25 @@ public static THttpService ofFormats(
...
 
-    THttpService(RpcService delegate, SerializationFormat[] supportedSerializationFormatArray) {
+    THttpService(RpcService delegate,
+                 SerializationFormat defaultSerializationFormat,
+                 Set<SerializationFormat> supportedSerializationFormats,
+                 BiFunction<ServiceRequestContext, ? super Throwable, ? extends Throwable> exceptionMapper) {
```
Comment (ikhoon) [inline thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java:274]: How about generalizing this more like `BiFunction<? super ServiceRequestContext, ? super Throwable, ? extends RpcResponse> exceptionHandler`?
I think we can use a similar approach we did in `GrpcServiceBuilder`
https://github.com/line/armeria/blob/480321970111ac6249775199fedadcc71b862e84/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java#L483-L500

Comment (minwoox) [reply thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java:274]: If we want to return `RpcResponse`, we have to add a decorator like we did in here. https://github.com/line/centraldogma/blob/master/server/src/main/java/com/linecorp/centraldogma/server/internal/thrift/CentralDogmaExceptionTranslator.java
So we can add a callback to propagate the event to here:
https://github.com/line/armeria/blob/master/thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java#L539
I'm not sure if it's worth.

File: thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java (hunk 1 above)
Comment (minwoox) [reply thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java]: Fixed. 😉

File: thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java (hunk 2 above)
Comment (minwoox) [reply thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpService.java]: Thanks, changed to return `RpcResponse`. 😉

File: thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpServiceBuilder.java [hunk 3]
```diff
@@ -61,30 +64,21 @@
  */
 public final class THttpServiceBuilder {
 
-    private static final SerializationFormat[] EMPTY_FORMATS = new SerializationFormat[0];
+    private static final BiFunction<? super ServiceRequestContext, ? super Throwable, ? extends RpcResponse>
+            default_exception_handler = (ctx, cause) -> RpcResponse.ofFailure(cause);
```
Comment (ikhoon) [inline thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpServiceBuilder.java:68]: UPPER_CASE or camelCase? 🤣

**Review by ikhoon — APPROVED** [review]
Thanks! @minwoox

File: thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpServiceBuilder.java (hunk 3 above)
Comment (minwoox) [reply thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpServiceBuilder.java:68]: 😱  😅

File: thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpServiceBuilder.java [hunk 4]
```diff
@@ -156,6 +150,17 @@ public THttpServiceBuilder defaultSerializationFormat(SerializationFormat defaul
         return this;
     }
 
+    /**
+     * Adds the {@link BiFunction} that returns an {@link RpcResponse} using the given {@link Throwable}
+     * and {@link ServiceRequestContext}.
```
Comment (trustin) [inline thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpServiceBuilder.java:155]: Shouldn't this start with `Sets`?

File: thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpServiceBuilder.java [hunk 5]
```diff
@@ -184,7 +189,21 @@ public THttpService build() {
...
+
+        return new THttpService(decorate(tcs), defaultSerializationFormat, builder.build(), exceptionHandler);
+    }
+
+    /**
+     * Returns a newly-created {@link THttpService} decorator with the properties of this builder.
+     */
+    public Function<? super RpcService, THttpService> newDecorator() {
+        return this::build0;
     }
```
Comment (trustin) [inline thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpServiceBuilder.java:208]: Could appear before `build0()` for readability?

File: thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpServiceBuilder.java (hunk 4 above)
Comment (minwoox) [reply thrift0.13/src/main/java/com/linecorp/armeria/server/thrift/THttpServiceBuilder.java]: Oops, fixed. 😉

**Review by trustin — APPROVED** [review]
Nice work, @minwoox :+1:

**Comment by minwoox** [comment]
Thanks, reviewers. 😄

**Pull request merged**